flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/FlinkRelOptUtil.scala

```diff
@@ -248,4 +251,262 @@ object FlinkRelOptUtil {
     }
   }
 
+  private[this] case class LimitPredicate(rankOnLeftSide: Boolean, pred: RexCall)
+
+  private[this] sealed trait Boundary
+
+  private[this] case class LowerBoundary(lower: Long) extends Boundary
+
+  private[this] case class UpperBoundary(upper: Long) extends Boundary
+
+  private[this] case class BothBoundary(lower: Long, upper: Long) extends Boundary
+
+  private[this] case class InputRefBoundary(inputFieldIndex: Int) extends Boundary
+
+  private[this] sealed trait BoundDefine
+
+  private[this] object Lower extends BoundDefine // defined lower bound
+  private[this] object Upper extends BoundDefine // defined upper bound
+  private[this] object Both extends BoundDefine // defined lower and uppper bound
+
+  /**
+    * Extracts the TopN offset and fetch bounds from a predicate.
+    *
+    * @param  predicate           predicate
+    * @param  rankFieldIndex      the index of rank field
+    * @param  rexBuilder          RexBuilder
+    * @param  config              TableConfig
+    * @return A Tuple2 of extracted rank range and remaining predicates.
+    */
+  def extractRankRange(
```

Comment:
Could you move these rank specific utilities to some `RankUtil`?

Reply:
OK